src/lsm/README.md

```diff
@@ -0,0 +1,111 @@
+# Tree
+## Compaction
+
+Tree compaction runs to the sound of music!
+
+Compacting LSM trees involves merging and moving tables into the next levels as needed.
+To avoid write amplification stalls and bound latency, compaction is done incrementally.
+
+A full compaction phase is denoted as a bar or measure, using terms from music notation.
+Each bar consists of `lsm_batch_multiple` beats or "compaction ticks" of work.
+A compaction beat is started asynchronously with `compact()` which takes a callback and
+the op that was just committed.
+
+A bar is split in half according to the "first" down beat and "middle" down beat.
+The first half of the bar compacts even levels while the latter compacts odd levels.
+Mutable table changes are sorted and compacted into the immutable table.
+The immutable table is compacted into level 0 during the odd level half of the bar.
+
+At any given point, there are at most levels/2 compactions running concurrently.
+The source level is denoted as `level_a` and the target level as `level_b`.
+The last level in the LSM tree has no target level so it is never a source level.
+
+Assuming a bar/`lsm_batch_multiple` of 4, the invariants can be described as follows:
+
+* At the end of every beat, there is space in mutable table for the next beat.
+* The manifest info for the compaction's input tables is updated at the end of the half-bar.
+* The manifest info for the compaction's output tables is updated during the compaction.
+* The manifest is compacted at the end of every beat.
+* The compactions' output tables are not visible until the compaction has finished.
+
+1. (first) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Allow level visible table counts to overflow if needed.
```

Comment:
What is a 'level visible table count' and why would it overflow?

Reply:
Is this more clear?:

>     * Allow the per-level visible table limits to overflow if needed.

Reply:
What are the visible table limits?

Reply:
I'll add these (in different sections):
```
- Level `0` … level `config.lsm_levels - 1` each contain an exponentially increasing number of
  on-disk tables: `config.lsm_growth_factor ^ (level + 1)`. (`config.lsm_growth_factor` is usually 8).
```

and:

```
    * Allow the per-level table limits to overflow if needed (for example, if we may compact a table
      from level `A` to level `B`, where level `B` is already full).
```